### PR TITLE
Refactor Event Listeners

### PR DESCRIPTION
**Pull request by wilhuff**
While working on iOS's event listeners I realized that the platforms have a bunch of spurious differences that don't really need to be there.

Highlights:
  * Rename ExecutorEventListener to AsyncEventListener to match iOS
  * Move ListenerRegistrationImpl and AsyncEventListener to core
  * Split ActivityScoping out of ListenerRegistrationImpl
  * Make the layering of ListenerRegistration clearer.

Note: this is running slightly in advance of forthcoming iOS changes

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentReference.java [hunk 1]
```diff
@@ -479,49 +481,58 @@ public ListenerRegistration addSnapshotListener(
...
+    AsyncEventListener<ViewSnapshot> asyncListener =
+        new AsyncEventListener<>(userExecutor, viewListener);
+
     com.google.firebase.firestore.core.Query query = asQuery();
-    QueryListener queryListener = firestore.getClient().listen(query, options, wrappedListener);
-    return new ListenerRegistrationImpl(
-        firestore.getClient(), queryListener, activity, wrappedListener);
+    QueryListener queryListener = firestore.getClient().listen(query, options, asyncListener);
+
+    return ActivityScope.bind(
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentReference.java]: Nifty.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 2]
```diff
@@ -891,30 +892,35 @@ public ListenerRegistration addSnapshotListener(
...
-              if (snapshot != null) {
-                QuerySnapshot querySnapshot = new QuerySnapshot(this, snapshot, firestore);
-                listener.onEvent(querySnapshot, null);
-              } else {
-                hardAssert(error != null, "Got event without value or error set");
-                listener.onEvent(null, error);
-              }
-            });
+      EventListener<QuerySnapshot> userListener) {
+
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java]:
```
// Convert from ViewSnapshots to QuerySnapshots.
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 3]
```diff
@@ -891,30 +892,35 @@ public ListenerRegistration addSnapshotListener(
...
+            userListener.onEvent(null, error);
+            return;
+          }
+
+          hardAssert(snapshot != null, "Got event without value or error set");
+
+          QuerySnapshot querySnapshot = new QuerySnapshot(this, snapshot, firestore);
+          userListener.onEvent(querySnapshot, null);
+        };
+
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java]:
```
// Call the viewListener on the userExecutor.
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/ListenerRegistrationImpl.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  private final FirestoreClient client;
+
+  /** The internal query listener object that is used to unlisten from the query. */
+  private final QueryListener queryListener;
+
+  /** The event listener for the query that raises events asynchronously. */
+  private final AsyncEventListener<ViewSnapshot> asyncEventListener;
+
+  /** Creates a new ListenerRegistration. Is activity-scoped if and only if activity is non-null. */
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/ListenerRegistrationImpl.java]: stale comment.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java (hunk 2 above)
Comment (wilhuff) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java]: Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java (hunk 3 above)
Comment (wilhuff) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java]: Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/ListenerRegistrationImpl.java (hunk 4 above)
Comment (wilhuff) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/ListenerRegistrationImpl.java]: Removed.

**Pull request merged**